### ISSUE TITLE
Collect2qBlocks introduces wrong dependencies

### ISSUE THREAD

**Issue by kdk** [Qiskit/qiskit-terra#2572]
```
>>> qc = qk.QuantumCircuit(2,1)
>>> qc.measure(1,0)
>>> qc.h(1)
>>> qc.cx(1,0)
>>> qk.execute(qc, qk.Aer.get_backend('qasm_simulator')).result().get_counts()
{'0': 1024}
>>> print(qc)
                ┌───┐
q_0: |0>────────┤ X ├
        ┌─┐┌───┐└─┬─┘
q_1: |0>┤M├┤ H ├──■──
        └╥┘└───┘
 c_0: 0 ═╩═══════════

>>> from qiskit.test.mock import FakeTenerife
>>> oqc = qk.transpile(qc, backend=FakeTenerife(), optimization_level=3)
>>> qk.execute(oqc, qk.Aer.get_backend('qasm_simulator')).result().get_counts()
{'0': 491, '1': 533}
>>> print(oqc)
                 ┌─────────────────────┐   ┌──────────────┐┌───┐┌──────────────┐┌───────────────────────────┐
      q_0: |0>───┤ U3(1.5708,5.9989,0) ├───┤ U2(0,3.1416) ├┤ X ├┤ U2(0,3.1416) ├┤ U3(1.5708,3.1416,-2.8573) ├───
              ┌──┴─────────────────────┴──┐├──────────────┤└─┬─┘├──────────────┤├───────────────────────────┤┌─┐
      q_1: |0>┤ U3(2.8573,-1.5708,1.5708) ├┤ U2(0,3.1416) ├──■──┤ U2(0,3.1416) ├┤ U3(1.5708,0.28425,3.1416) ├┤M├
              └───────────────────────────┘└──────────────┘     └──────────────┘└───────────────────────────┘└╥┘
ancilla_0: |0>────────────────────────────────────────────────────────────────────────────────────────────────╫─
                                                                                                              ║
ancilla_1: |0>────────────────────────────────────────────────────────────────────────────────────────────────╫─
                                                                                                              ║
ancilla_2: |0>────────────────────────────────────────────────────────────────────────────────────────────────╫─
                                                                                                              ║
       c_0: 0 ════════════════════════════════════════════════════════════════════════════════════════════════╩═
```

**Comment by mtreinish** [comment]
I added a circuit print to each stage in the passmanager (which might be a handy debug flag to add in the future) to see where this is being moved, it looks like one of the consolidate blocks passes is where it moves while doing the fixed depth loop. It goes from:
```
<qiskit.transpiler.passes.collect_2q_blocks.Collect2qBlocks object at 0x7f0fb5aee9b0>
              ┌────────────────────┐   ┌──────────────┐                  ┌───┐┌──────────────┐┌───────────────────────────────┐
      q_0: |0>┤ U2(-1.5708,3.1416) ├───┤ U2(0,3.1416) ├──────────────────┤ X ├┤ U2(0,3.1416) ├┤ U3(1.5708,3.3307e-16,-1.5708) ├
              └────────┬─┬─────────┘┌──┴──────────────┴─┐┌──────────────┐└─┬─┘├──────────────┤└────┬─────────────────────┬────┘
      q_1: |0>─────────┤M├──────────┤ U2(4.7124,1.5708) ├┤ U2(0,3.1416) ├──■──┤ U2(0,3.1416) ├─────┤ U2(-1.5708,-3.1416) ├─────
                       └╥┘          └───────────────────┘└──────────────┘     └──────────────┘     └─────────────────────┘     
ancilla_0: |0>──────────╫──────────────────────────────────────────────────────────────────────────────────────────────────────
                        ║                                                                                                      
ancilla_1: |0>──────────╫──────────────────────────────────────────────────────────────────────────────────────────────────────
                        ║                                                                                                      
ancilla_2: |0>──────────╫──────────────────────────────────────────────────────────────────────────────────────────────────────
                        ║                                                                                                      
       c_0: 0 ══════════╩══════════════════════════════════════════════════════════════════════════════════════════════════════
                                                                                                   
```
(which remains unchanged from the cxdirection pass earlier in that loop iteration)
to:

```
<qiskit.transpiler.passes.consolidate_blocks.ConsolidateBlocks object at 0x7f0fb5aee9e8>
              ┌──────────┐   
      q_0: |0>┤0         ├───
              │  unitary │┌─┐
      q_1: |0>┤1         ├┤M├
              └──────────┘└╥┘
ancilla_0: |0>─────────────╫─
                           ║ 
ancilla_1: |0>─────────────╫─
                           ║ 
ancilla_2: |0>─────────────╫─
                           ║ 
       c_0: 0 ═════════════╩═

```

**Comment by nonhermitian** [comment]
So for the actual devices, the measurements have to occur at the end.  So, while there is a bug here, depending on how literally you take "transpiling targets a given device", this example should throw an exception to begin with since the input circuit is invalid for the target backend.

**Comment by maddy-tod** [comment]
Running this again it now returns
```
                ┌─┐            ┌───┐┌──────────┐
      (q1) q0|0>┤M├────────────┤ X ├┤ U2(0,pi) ├
                └╥┘┌──────────┐└─┬─┘├──────────┤
      (q0) q1|0>─╫─┤ U2(0,pi) ├──■──┤ U2(0,pi) ├
                 ║ └──────────┘     └──────────┘
(ancilla0) q2|0>─╫──────────────────────────────
                 ║                              
(ancilla1) q3|0>─╫──────────────────────────────
                 ║                              
(ancilla2) q4|0>─╫──────────────────────────────
                 ║                              
         c_0: 0 ═╩══════════════════════════════
```
Which I think is the expected output with the measure at the front. This was fixed by #2764

**Comment by ajavadia** [comment]
^ Yeah that looks right -- closing.
